### ISSUE TITLE
DashboardScreen is refreshed every time tracked categories are added/removed/updated

### ISSUE THREAD

**Issue by Doan-Pham** [Doan-Pham/Stren#15]
### TLDR
DashboardViewModel uses MutableStateFlow in multiple places to handle data for use in DashboardViewModel. MutableStateFlow is updated only if its value is assigned to a whole new object -> any small update (Ex: update to only 1 or 2 fields) to MutableStateFlow requires creating a whole new object with the updated bits -> whole new object = all data is reset -> screen is refreshed since all of its data changed

### Current implementation for DashboardViewModel and why it causes the issue
The current data flow for DashboardViewModel and Dashboard is:
**(1)** **[GetUserFullDataUseCase]**: Exposes a Flow of user's tracked categories in a Flow\<TrackedCategory\>
**(2)** **[DashboardViewModel]**: flatMapLatest the flow from (1) and create **StateFlow<List<MutableStateFLow\<DataOutput\>>>**

- a. Outermost **StateFlow** - To use collectAsStateWithLifeCycle() in **(3)**, or else, it'll be difficult to receive real-time update from **(1)** and map to data
- b. **List<MutableStateFlow\<DataOutput\>>** - since each DataOuput is created from a separate repository, and individual update to any single MutableStateFlow\<DataOuput\> shouldn't affect other MutableStateFlow\<DataOuput\>

      
**(3)** **[DashboardScreen ]**: collect the  StateFlow<List<MutableStateFLow\<DataOutput\>>>  from (2) and updates UI

Since MutableStateFlow is an observable data holder that is only updated if its value is assigned to a whole new object, any change (update/add/remove) to the Flow from **(1)** forces creating a whole new value for StateFlow in **(2)** which in turns reset all data collected in **(3)**

### Alternative
The 1st implementation for DashboardViewModel uses Jetpack Compose's SnapshotStateList and the data class "DataFetchingTrigger". 
- Jetpack Compose's **SnapshotStateList** works well will lists and don't require resetting the whole list to add/remove/update data. So the  **StateFlow<List**<MutableStateFLow\<DataOutput\>>> from **(2)** previous section becomes **SnapshotStateList**<MutableStateFLow\<DataOutput\>>
- **DataFetchingTrigger** data class has the same properties as the TrackedCategory in the flow in **(1)** from previous section. This data class basically acts as a cache for data returned from **(1)** and any updates from user input actually happen on this cache, allowing for localizing changes and avoiding resetting all data

Although this solution works well for the case of updating an element in DashboardScreen's list, it's not enough for adding/removing. The cause is that in DashboardScreen, every MutableStateFlow\<DataOuput\> from DashboardViewModel is collected and then added to a List<State> which resets itself on every recomposition. This was supposed to be solved using a combination of remember{}  and adding only the MutableStateFlow that's not already collected in List<State>, but for some reason, List<State> will stop receiving updates from MutableStateFlow once it's added. 

For the above reason, coupled with the fact that cache invalidation is complex, this alternative was abandoned in favor of the current solution

 ### TODO
Probably bring back the cache from the alternative solution and try to deal with its shortcoming
